File: 源码仓库/backend/app/schemas/task.py
"""标注任务 + 标注结果的请求/响应模型"""
from datetime import datetime
from pydantic import BaseModel


# ── 任务 ──

class LabelTaskCreate(BaseModel):
    """创建任务——owner_id 由后端从 JWT 获取，前端无需传递"""
    title: str
    description: str = ""
    tags: str = ""
    dataset_id: int
    schema_id: int
    assignee_id: int | None = None                                   # 指派模式指定标注员，抢单留空
    assignee_type: str = "labeler"
    deadline: datetime | None = None
    quota: int = 0
    reward_per_item: float | None = None
    reward_cap: float | None = None
    distribution_strategy: str = "first_come"
    grab_limit: int | None = None
    ai_agent_id: int | None = None

    model_config = {"json_schema_extra": {"examples": [{"title": "文本情感分类标注任务", "description": "对用户评论进行情感极性标注", "tags": "情感分析,NLP", "dataset_id": 1, "schema_id": 1, "quota": 100, "distribution_strategy": "first_come", "reward_per_item": 0.5, "reward_cap": 200.0, "deadline": "2026-12-31T23:59:59"}]}}


class LabelTaskUpdate(BaseModel):
    """更新任务——所有字段可选"""
    title: str | None = None
    description: str | None = None
    tags: str | None = None
    assignee_id: int | None = None
    assignee_type: str | None = None
    deadline: datetime | None = None
    quota: int | None = None
    reward_per_item: float | None = None
    reward_cap: float | None = None
    distribution_strategy: str | None = None
    status: str | None = None
    ai_agent_id: int | None = None


class LabelTaskResponse(BaseModel):
    """返回任务详情——progress 由 Model 属性自动计算"""
    id: int
    title: str
    description: str
    tags: str
    dataset_id: int
    schema_id: int
    ai_agent_id: int | None
    owner_id: int
    assignee_id: int | None
    assignee_type: str
    deadline: datetime | None
    quota: int
    reward_per_item: float | None
    reward_cap: float | None
    distribution_strategy: str
    grab_limit: int | None = None
    status: str
    total_items: int
    completed_items: int
    progress: float
    claimed_items: int = 0
    schema_snapshot: dict | None = None
    schema_version: int = 1
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── 标注结果 ──

class LabelResultCreate(BaseModel):
    """提交标注结果——task_id 由路径参数提供，labeler_id 从 JWT 获取"""
    item_id: int
    data: dict
    round: int = 1

    model_config = {"json_schema_extra": {"examples": [{"item_id": 1, "data": {"category": "科技", "sentiment": "positive", "keywords": ["AI", "芯片"]}, "round": 1}]}}


class LabelResultUpdate(BaseModel):
    """审核更新——审核员填写状态、意见、AI 评分"""
    status: str | None = None
    comment: str | None = None
    reviewer_id: int | None = None
    reviewed_at: datetime | None = None
    ai_scores: dict | None = None


class LabelResultResponse(BaseModel):
    """返回标注结果详情——含审核追溯信息"""
    id: int
    task_id: int
    item_id: int
    labeler_id: int
    labeler_type: str
    data: dict
    round: int
    status: str
    comment: str
    reviewer_id: int | None
    reviewed_at: datetime | None
    ai_scores: dict | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── 标注工作台：任务条目（task_items JOIN dataset_items）──

class TaskItemResponse(BaseModel):
    """左侧队列——每一条目的原始数据 + 本任务内状态"""
    id: int                                                     # dataset_item.id
    index: int
    data: dict
    status: str                                                 # pending / labeled / skipped
    labeler_id: int | None


class TaskItemDetail(BaseModel):
    """中间表单区——单条详情 + 该标注员最近一次提交（被打回时展示驳回理由）"""
    id: int
    index: int
    data: dict
    status: str
    labeler_id: int | None
    last_result: LabelResultResponse | None = None


class LabelerStats(BaseModel):
    """右侧统计卡片"""
    pending: int
    labeled: int
    skipped: int
    total: int

File: 源码仓库/frontend/src/types/models/task.ts
/**
 * task.ts — 标注任务相关类型定义
 * Author: hongchuwudi
 * Description: 定义 LabelTask / LabelResult 等前端 camelCase 类型及后端 API 原始 snake_case 类型
 */

/** 标注任务（前端 camelCase 版） */
export interface LabelTask {
  id: number
  title: string
  description: string
  tags: string
  datasetId: number
  schemaId: number
  ownerId: number
  assigneeId: number | null
  assigneeType: 'labeler' | 'ai_agent'
  deadline: string | null
  quota: number
  distributionStrategy: 'first_come' | 'assigned' | 'quota_grab'
  status: 'draft' | 'published' | 'paused' | 'ended'
  totalItems: number
  completedItems: number
  progress: number
  createdAt: string
  updatedAt: string
}

/** 标注结果——单条数据的标注记录（前端 camelCase 版） */
export interface LabelResult {
  id: number
  taskId: number
  itemId: number
  labelerId: number
  labelerType: 'human' | 'ai'
  data: Record<string, unknown>
  round: number
  status: 'submitted' | 'ai_reviewing' | 'review' | 'final_review' | 'warehouse' | 'rejected'
  comment: string
  reviewerId: number | null
  reviewedAt: string | null
  aiScores: Record<string, number> | null
  createdAt: string
}

// ── 以下为后端 API 原始 snake_case 类型 ──

/** 任务列表项（API 原始格式） */
export interface TaskListItem {
  id: number
  title: string
  description: string
  tags: string
  status: string
  dataset_id: number
  schema_id: number
  ai_agent_id: number | null
  owner_id: number
  assignee_id: number | null
  assignee_type: string
  distribution_strategy: string
  grab_limit: number | null
  quota: number
  reward_per_item: number | null
  reward_cap: number | null
  total_items: number
  completed_items: number
  claimed_items: number
  progress: number
  deadline: string | null
  schema_snapshot: Record<string, unknown> | null
  schema_version: number
  created_at: string
  updated_at: string
}

/** 任务详情 */
export type TaskDetail = TaskListItem

/** 创建任务请求 */
export interface CreateTaskData {
  title: string
  description?: string
  tags?: string
  dataset_id: number
  schema_id: number
  assignee_id?: number | null
  distribution_strategy?: string
  grab_limit?: number | null
  quota?: number
  reward_per_item?: number | null
  reward_cap?: number | null
  deadline?: string | null
}

/** 更新任务请求 */
export interface UpdateTaskData {
  title?: string
  description?: string
  tags?: string
  dataset_id?: number
  schema_id?: number
  ai_agent_id?: number | null
  assignee_id?: number | null
  assignee_type?: string
  status?: string
  quota?: number
  reward_per_item?: number | null
  reward_cap?: number | null
  deadline?: string | null
  distribution_strategy?: string
  grab_limit?: number | null
}

/** 提交标注结果请求 */
export interface SubmitResultData {
  item_id: number
  data: Record<string, unknown>
  round?: number
}

/** 标注任务条目（API 原始格式） */
export interface TaskItemResponse {
  id: number
  index: number
  data: Record<string, unknown>
  status: 'pending' | 'labeled' | 'skipped'
  flow_history: FlowEntry[]
  labeler_id: number | null
  last_result?: {
    id: number
    task_id: number
    item_id: number
    labeler_id: number
    data: Record<string, unknown>
    round: number
    status: string
    comment: string
    ai_scores?: {
      verdict: string
      summary: string
      dimensions: { name: string; score: number; reason?: string }[]
      overall_score: number
    } | null
  } | null
}

/** 流转历史条目 */
export interface FlowEntry {
  status: string
  time: string
  actor: string
  actor_name: string
  round: number
  detail: string
}

/** 标注任务条目详情 */
export interface TaskItemDetail extends TaskItemResponse {
  last_result: {
    id: number
    task_id: number
    item_id: number
    labeler_id: number
    labeler_type: string
    data: Record<string, unknown>
    round: number
    status: 'submitted' | 'ai_reviewing' | 'review' | 'final_review' | 'warehouse' | 'rejected'
    comment: string
    reviewer_id: number | null
    reviewed_at: string | null
    ai_scores: {
      verdict: string
      summary: string
      dimensions: { name: string; score: number; reason?: string }[]
      overall_score: number
    } | null
    created_at: string
  } | null
}

/** 标注员统计数据 */
export interface LabelerStats {
  pending: number
  in_review: number
  rejected: number
  done: number
  total: number
}

/** 标注结果条目（API 原始格式） */
export interface LabelResultItem {
  id: number
  task_id: number
  item_id: number
  labeler_id: number
  labeler_type: string
  data: Record<string, unknown>
  round: number
  status: 'submitted' | 'ai_reviewing' | 'review' | 'final_review' | 'warehouse' | 'rejected'
  comment: string
  reviewer_id: number | null
  reviewed_at: string | null
  ai_scores: Record<string, unknown> | null
  created_at: string
}

/** 审计日志条目 */
export interface AuditLogEntry {
  id: number
  actor_id: number
  actor_name: string
  actor_role: string
  entity_type: string
  entity_id: number
  task_id: number
  action: string
  from_status: string
  to_status: string
  detail: string
  created_at: string
}

/** 导出任务 */
export interface ExportJob {
  id: number
  task_id: number
  format: string
  status: 'pending' | 'processing' | 'done' | 'failed'
  file_name: string
  item_count: number
  field_mapping: string | null
  error_message: string | null
  created_at: string
  finished_at: string | null
}
